Assets/MRTK/Providers/LeapMotion/LeapMotionArticulatedHand.cs

```diff
@@ -0,0 +1,299 @@
+                foreach (TrackedHandJoint joint in Enum.GetValues(typeof(TrackedHandJoint)))
+                {
+                    AttachmentHands.attachmentPoints |= ConvertMRTKJointToLeapJoint(joint);
+                }            
+            }
+        }
+#endif
+        public override bool TryGetJoint(TrackedHandJoint joint, out MixedRealityPose pose)
+        {
+#if LEAPMOTIONCORE_PRESENT
+            if (attachmentHand != null && attachmentHand.isTracked)
+            {
+                // Is the current joint a metacarpal
+                bool isMetacarpal = Array.Exists(metacarpals, (metacarpalJoint => metacarpalJoint == joint));
+
+                // AttachmentPointFlags does not include metacarpals.  
+                if (isMetacarpal) 
+                {
+                    MixedRealityPose metacarpalPose = GetMetacarpals(joint);
+
+                    pose = metacarpalPose;
+                    return true;
+                }
+
+                AttachmentPointFlags leapAttachmentFlag = ConvertMRTKJointToLeapJoint(joint);
+
+                // Get the pose of the leap joint
+                AttachmentPointBehaviour leapJoint = attachmentHand.GetBehaviourForPoint(leapAttachmentFlag);
+
+                // Set the pose calculated by the leap motion to a mixed reality pose
+                pose = new MixedRealityPose(leapJoint.transform.position, leapJoint.transform.rotation);
+                return true;
+            }          
+#endif
+            pose = MixedRealityPose.ZeroIdentity;
+            return false;
+        }
+#if LEAPMOTIONCORE_PRESENT
+
+        // Convert leap hand chirality to MRTK handedness
+        private Handedness HandChiralityToHandedness(Chirality chirality)
+        {
+            return (chirality == Chirality.Left) ? Handedness.Left : Handedness.Right;
+        }
+
+        /// <summary>
+        /// Get the pose of the metacarpal joints from the current frame of the LeapServiceProvider because the metacarpal joints 
+        /// are not included in AttachmentPointFlags.  The metacarpal joints are those located directly above the wrist.
+        /// </summary>
+        /// <param name="metacarpalJoint">A metacarpal TrackedHandJoint</param>
+        /// <returns>The MixedRealityPose for the leap metacarpal joint</returns>
+        private MixedRealityPose GetMetacarpals(TrackedHandJoint metacarpalJoint)
+        {
+            // Get the leap hand data in the current frame
+            List<Hand> hands = LeapMotionDeviceManager.CurrentHandsDetectedByLeap;
+
+            Hand leapHand;
+
+            if (ControllerHandedness == Handedness.Left)
+            {
+                leapHand = hands.Find(hand => hand.IsLeft);
+            }
+            else
+            {
+                leapHand = hands.Find(hand => hand.IsRight);
+            }
+
+            int metacarpalIndex = Array.IndexOf(metacarpals, metacarpalJoint);
```

Comment:
Is there any way we can reduce the number of array finds (i.e. indexOf, Find) - if we're gonna be using an enum like TrackedHandJoint we can always:

1) Use a set
2) Just use the int value of the hand joint as the array index.

Reply:
I changed the metacarpals Array to a list to remove the Array.Finds, but indexOf is still needed in this case.  The metacarpals list is in a specific order and the index is needed because it maps each MRTK metacarpal joint to the correct metacarpal joint in the leap hand.  I also cannot use the int value of a metacarpal TrackedHandJoint because the value does not map to the Leap metacarpal joints.

```
        private MixedRealityPose GetMetacarpalPose(TrackedHandJoint metacarpalJoint)
        {
            int metacarpalIndex = metacarpals.IndexOf(metacarpalJoint);

            // Get the pose of the metacarpal 
            Vector3 position = leapMotionMetacarpalHand.Fingers[metacarpalIndex].bones[0].PrevJoint.ToVector3();
            Quaternion rotation = leapMotionMetacarpalHand.Fingers[metacarpalIndex].bones[0].Rotation.ToQuaternion();

            return new MixedRealityPose(position, rotation);
        }
```